modules/cfe_testcase/src/es_counter_test.c

```diff
@@ -0,0 +1,166 @@
+**
+**      Licensed under the Apache License, Version 2.0 (the "License");
+**      you may not use this file except in compliance with the License.
+**      You may obtain a copy of the License at
+**
+**        http://www.apache.org/licenses/LICENSE-2.0
+**
+**      Unless required by applicable law or agreed to in writing, software
+**      distributed under the License is distributed on an "AS IS" BASIS,
+**      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+**      See the License for the specific language governing permissions and
+**      limitations under the License.
+**
+*************************************************************************/
+
+/**
+ * @file
+ *
+ * Functional test of ES Generic Counter APIs
+ */
+
+#include "cfe_test.h"
+
+void TestCounterCreateDelete(void)
+{
+    CFE_ES_CounterId_t Ids[CFE_PLATFORM_ES_MAX_GEN_COUNTERS];
+    CFE_ES_CounterId_t TestId;
+    CFE_ES_CounterId_t CheckId;
+    char               CounterName[CFE_MISSION_MAX_API_LEN];
+    char               CheckName[CFE_MISSION_MAX_API_LEN];
+    CFE_Status_t       Status;
+    uint32             NumCounters;
+    uint32             Idx;
+
+    UtPrintf("Testing: CFE_ES_RegisterGenCounter");
+
+    snprintf(CounterName, sizeof(CounterName), "ut");
+
+    /* Confirm proper bad argument rejection */
+    UtAssert_INT32_EQ(CFE_ES_RegisterGenCounter(&Ids[0], NULL), CFE_ES_BAD_ARGUMENT);
+    UtAssert_INT32_EQ(CFE_ES_RegisterGenCounter(NULL, CounterName), CFE_ES_BAD_ARGUMENT);
+
+    /* Create up to CFE_PLATFORM_ES_MAX_GEN_COUNTERS and confirm success */
+    for (NumCounters = 0; NumCounters < CFE_PLATFORM_ES_MAX_GEN_COUNTERS; ++NumCounters)
+    {
+        snprintf(CounterName, sizeof(CounterName), "C%u", (unsigned int)NumCounters);
+        Status = CFE_ES_RegisterGenCounter(&Ids[NumCounters], CounterName);
+        if (Status != CFE_SUCCESS)
+        {
+            break;
+        }
+    }
+
+    /* Confirm that the expected number of counters were created */
+    UtAssert_UINT32_EQ(NumCounters, CFE_PLATFORM_ES_MAX_GEN_COUNTERS);
+
+    /* Attempt to create one too many */
+    snprintf(CounterName, sizeof(CounterName), "extra");
+    UtAssert_INT32_EQ(CFE_ES_RegisterGenCounter(&TestId, CounterName), CFE_ES_NO_RESOURCE_IDS_AVAILABLE);
+
+    /* pick a single counter ID from the middle of the set for more detail testing of support APIs */
+    TestId = Ids[NumCounters / 2];
+    snprintf(CounterName, sizeof(CounterName), "C%u", (unsigned int)NumCounters / 2);
+
+    UtPrintf("Testing: CFE_ES_CounterID_ToIndex");
+
+    /* Confirm CFE_ES_CounterID_ToIndex works (nominal) */
+    Idx = UINT32_MAX;
+    UtAssert_INT32_EQ(CFE_ES_CounterID_ToIndex(TestId, &Idx), CFE_SUCCESS);
+    UtAssert_UINT32_LT(Idx, CFE_PLATFORM_ES_MAX_GEN_COUNTERS);
+
+    /* Confirm proper rejection of bad args in CFE_ES_CounterID_ToIndex */
+    UtAssert_INT32_EQ(CFE_ES_CounterID_ToIndex(CFE_ES_COUNTERID_UNDEFINED, &Idx), CFE_ES_ERR_RESOURCEID_NOT_VALID);
```

Comment:
CFE_ES_CounterID_ToIndex is being tested by https://github.com/nasa/cFE/pull/1716 so you don't need to check it here.

Reply:
Yeah, we wrote these tests in parallel.  I don't see a real problem with testing it in two places, as they are slightly different variations.  It doesn't hurt anything.

Reply:
alright just wanted to make sure we were aware.